### DISCUSSION TITLE
tracing-opentelemetry: why is `SpanBuilder::start` only called in `on_close` instead of `new_span`?

### DISCUSSION THREAD

**Discussion by LukeMathWalker** [tokio-rs/tracing#2133] [Q&A] [answered]
This is not really a feature request, but more of an attempt to understand the design decisions taken in `tracing-opentelemetry`.
When `new_span` gets invoked, it assembles a `SpanBuilder` which is injected into the `tracing`'s `Span` context (wrapped in `OtelData`) and manipulated throughout the lifecycle of the corresponding `tracing`'s Span. We only build an `opentelemetry::Span` in `on_close`, using `SpanBuilder::start_with_context`.

Am I correct in assuming that this means that `Context::current().span()` will not return the OTEL counterpart to the currently active `opentelemetry::Span` given that we do not call `start_with_context` until the `on_close` callback is invoked (i.e. the corresponding `tracing::Span` is no longer active)?

**Comment by jtescher** [comment, marked as answer]
You are correct that otel's global `Context` is not used to track current spans as they are instead paired with tracing spans and tracked through the tracing registry. The main reason that an otel span is not started until the associated tracing span ends is to support setting a parent trace id / context _after_ the associated tracing span has been created, which the otel API does not support.

Cases where you would want `Context::current().span()` are generally handled via either tracing's `Span::current()` for setting attributes, and [`OpenTelemetrySpanExt::context`](https://docs.rs/tracing-opentelemetry/latest/tracing_opentelemetry/trait.OpenTelemetrySpanExt.html#tymethod.context) to read / inject the current context into a client request or otherwise bridge back to otel.

**Comment by LukeMathWalker** [comment]
Thanks for clarifying!
I assume the reason for `set_parent` is that there is no way to do that via a "special" field when the span is declared, as you do for the span name and other OTEL-specific information.

**Reply by jtescher** [reply]
Yep, exactly.
